### PR TITLE
Extensions: Add block editor (Gutenberg) extensions source

### PR DESCRIPTION
**Pull request by roccotripaldi**
Fixes nothing, aside from developer headaches

#### Changes proposed in this Pull Request:

This PR copies the javascript from the wpcalypso repo, and into Jetpack, with the hopes of building the blocks from here in Jetpack, where their PHP code lives.

#### Testing instructions:

- Does the file structure make sense?
- Try building:
  1. copy `extensions/blocks` folder into your local calypso at `packages/jetpack-blocks/src/blocks`
  1. In `packages/jetpack-blocks/src/blocks` run `npm install`
  1. In calypso root run `npx lerna run build --stream --scope='@automattic/jetpack-blocks' -- -- --output-path /path-to-jetpack/_inc/blocks`
  1. Does it build ok?

#### Proposed changelog entry for your changes:

* Adding source code for Jetpack blocks

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
Not to throw a wrench into this or anything, but in #11312 we had discussed a somewhat different file organization: 

https://github.com/Automattic/jetpack/pull/11312/commits/12fa4fdf24e43e9bfec8850386f6948099050a80#diff-bf97bdc65b5272e6bbde848e059bdd4eR14

I'm fine either way but wanted to point this out. Wouldn't it be cleaner / easier to handle/build to have the jsx in their own folders?Not to throw a wrench into this or anything, but in #11312 we had discussed a somewhat different file organization: 

https://github.com/Automattic/jetpack/pull/11312/commits/12fa4fdf24e43e9bfec8850386f6948099050a80#diff-bf97bdc65b5272e6bbde848e059bdd4eR14

I'm fine either way but wanted to point this out. Wouldn't it be cleaner / easier to handle/build to have the jsx in their own folders?

**Comment by simison** [comment]
@jeherve @lezama @sirreal and I discussed folder structure and concluded to keep it as flat (opposed to splitting js+scss files to `client/` and php files to `backend/` folders.)

Gutenberg also has flat structure:

https://github.com/WordPress/gutenberg/tree/95642a40ca5dddbaf93a565d9a78a554974a3e0e/packages/block-library/src/search

Concern was that it would be confusing for complicated blocks with several files and that's also how Jetpack Dashboard files are split.

We can iterate once we _do_ have more files/complex blocks (instead of optimizing early).

One aspect was to try match "single block plugin" proposal (https://make.wordpress.org/meta/2019/03/08/the-block-directory-and-a-new-type-of-plugin) but that's also something we might not want to optimize too early for, just keep it in mind:
> To begin with, Single Block Plugins will contain a single PHP file containing plugin headers and initialization code only. Eventually, this should become unnecessary.

**Comment by simison** [comment]
This depends on including files to ignores in different places before this is merged: https://github.com/Automattic/jetpack/pull/11639

**Comment by simison** [comment]
We should update docs file: [`extensions/README.md`](https://github.com/Automattic/jetpack/blob/36a8543b9933b9d4a287f5f7ad7f2dc45bed1389/extensions/README.md)

Maybe something from these files:
- https://github.com/Automattic/wp-calypso/tree/cd16c9b3d48cbfdf370d000e65be315c566cd5f4/packages/jetpack-blocks#readme
- https://github.com/Automattic/wp-calypso/tree/cd16c9b3d48cbfdf370d000e65be315c566cd5f4/client/gutenberg/extensions#readme

File: extensions/README.md [hunk 1]
```diff
@@ -1,40 +1,37 @@
...
 ## Extension Structure
 
-Your extension should follow this structure:
+Extensions loosely follow this structure:
 
 ```
 .
-└── blockname/
-	└── blockname.php ← PHP file where the block and its assets are registered.
+└── block-or-plugin-name/
```
Comment (simison) [inline extensions/README.md]: In https://github.com/Automattic/jetpack/pull/11333 (unmerged) @ockham suggested separate folder for Plugins: `extensions/plugins/seo/seo.php`

Thoughts?

Comment (simison) [reply extensions/README.md]: Suppose extension could be both at the same time so it might not make sense to optimize this way?

**Comment by roccotripaldi** [comment]
We can ignore the failing wpcom sync and manually sync the file in a different phabricator.

**Comment by kraftbj** [comment]
Assigning to @jeherve for the review, scheduled for review today.

File: extensions/README.md (hunk 1 above)
Comment (jeherve) [reply extensions/README.md]: It may be easier for our build tools and for new contributors if everything was in one place?

**Review by jeherve — COMMENTED** [review]
The blocks seem to build nicely. I do have some questions though.

Should the imported blocks be free of eslint errors? I would assume so since we have Prettier running in Calypso. I can currently see errors like this one, in `extensions/shared/frontend-management.js`:

```
Import individual methods from the Lodash module.
```

Since switching from jsx to js, we also have a few errors like this one, in `extensions/shared/help-message.js` or in `extensions/shared/jetpack-plugin-sidebar.js` for example:

```
'React' must be in scope when using JSX
```

**Comment by simison** [comment]
We'll need to bring this change here when the PR merges: https://github.com/Automattic/wp-calypso/pull/31690

File: extensions/README.md (hunk 1 above)
Comment (simison) [reply extensions/README.md]: Does `/extensions/blocks` then make sense and should we just have `/extensions` ?

Comment (sirreal) [reply extensions/README.md]: I believe the things in `/blocks` are actually blocks.

#11640 includes additional extensions (the sidebar and category) that aren't blocks. They are not located under blocks.

**Comment by jeherve** [comment]
I would be tempted to close this PR in favor of #11639 now that the commits are included in that other PR as well, but there are still some questions above that I think would be worth discussing.

**Comment by sirreal** [comment]
I've pulled the jsx -> js rename commit 2cdb5b28ee871b1da7495b555f6bc58726eafa96 over from #11639

**Comment by sirreal** [comment]
> I would be tempted to close this PR in favor of #11639 now that the commits are included in that other PR as well, but there are still some questions above that I think would be worth discussing.

I've cleaned up #11639 significantly so these should be independent again.

**Review by ockham — DISMISSED** [review]
This seems to be working well for me. Let's get #11639 merged and this one soon afterwards to reduce the duplication (and confusion) for developers. I think it's fine to address lint errors etc after the fact.

**Comment by sirreal** [comment]
I've just updated and squashed this branch to include the latest Calypso master source (as of https://github.com/Automattic/wp-calypso/commit/0f0ba7ba6386c78b888aae7755b7d7de4e6114a4).

Ready for review.

**Review by kraftbj — APPROVED** [review]

**Review by lezama — APPROVED** [review]
:partyparrot: :shipit: 🚢 IT

**Pull request merged**